src/evm/protocol/uniswap_v4/hooks/generic_vm_hook_handler.rs

```diff
@@ -152,11 +151,52 @@ where
 
     fn after_swap(
         &self,
-        _params: AfterSwapParameters,
+        params: AfterSwapParameters,
+        block: u64,
     ) -> Result<WithGasEstimate<BeforeSwapDelta>, SimulationError> {
-        self.unlock_pool_manager();
-        todo!()
-        // self.contract.call(..)
+        let transient_storage_params = self.unlock_pool_manager();
+
+        let args = (
+            params.sender,
+            (
+                params.context.currency_0,
+                params.context.currency_1,
+                Uint::<24, 1>::from(params.context.fees.lp_fee),
+                Signed::<24, 1>::try_from(params.context.tick).map_err(|e| {
+                    SimulationError::FatalError(format!("Failed to convert tick: {e:?}"))
+                })?,
+                self.address,
+            ),
+            (
+                params.swap_params.zero_for_one,
+                params.swap_params.amount_specified,
+                params.swap_params.sqrt_price_limit,
+            ),
+            params.delta,
+            params.hook_data.to_vec(),
+        );
+        let selector = "afterSwap(address,(address,address,uint24,int24,address),(bool,int256,uint160),int256,bytes)";
+
+        let res = self.contract.call(
+            selector,
+            args,
+            block,
+            None,
+            None,
+            Some(self.pool_manager),
+            U256::from(0u64),
+            Some(transient_storage_params),
+        )?;
+
+        let decoded = AfterSwapReturn::abi_decode(&res.return_value).map_err(|e| {
+            SimulationError::FatalError(format!("Failed to decode before swap return value: {e:?}"))
+        })?;
+        Ok(WithGasEstimate {
```

Comment:
Should we add a TODO to verify the problem that we encountered with transient storage in other contracts? Or open another task for it?